### PR TITLE
chore: Added NumaflowController Unit test for PPND, updated ISBService Unit test for PPND

### LINKED ISSUE #168: Unit testing for Pause feature
# Summary

Add more unit testing, such as for `reconcile()` functions

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by juliev0**
Fixes #168 

### Modifications

Added a unit test for the `NumaflowControllerRollout` Controller reconcile() method for PPND cases.
Made some updates to test the `ISBServiceRollout` Controller reconcile() method, such as setting and checking its in-memory pause request setting.

Consolidated some shared logic into a common file.

### CODE REVIEW

File: internal/controller/isbservicerollout_controller.go [hunk 1]
```diff
@@ -402,7 +402,8 @@ func (r *ISBServiceRolloutReconciler) isISBServiceUpdating(ctx context.Context,
 
 func (r *ISBServiceRolloutReconciler) getPipelineList(ctx context.Context, rolloutNamespace string, rolloutName string) ([]*kubernetes.GenericObject, error) {
 	// here
-	return kubernetes.ListCR(ctx, r.restConfig, common.NumaflowAPIGroup, common.NumaflowAPIVersion, "pipelines", rolloutNamespace, fmt.Sprintf("%s=%s,%s", common.LabelKeyISBServiceNameForPipeline, rolloutName, common.LabelKeyPipelineRolloutForPipeline), "")
+	return kubernetes.ListCR(ctx, r.restConfig, common.NumaflowAPIGroup, common.NumaflowAPIVersion, "pipelines", rolloutNamespace,
+		fmt.Sprintf("%s=%s,%s", common.LabelKeyISBServiceNameForPipeline, rolloutName, common.LabelKeyPipelineRolloutForPipeline), "")
```
Comment (juliev0) [inline internal/controller/isbservicerollout_controller.go:406]: this is just a new line

File: internal/controller/common_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	numaflowv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	numaflowversioned "github.com/numaproj/numaflow/pkg/client/clientset/versioned"
+	"github.com/numaproj/numaplane/internal/common"
+	apiv1 "github.com/numaproj/numaplane/pkg/apis/numaplane/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclientgo "k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
```
Comment (juliev0) [inline internal/controller/common_test.go:17]: moved some common code from the tests over here

File: internal/controller/isbservicerollout_controller_test.go [hunk 3]
```diff
@@ -268,107 +273,139 @@ func Test_reconcile_isbservicerollout_PPND(t *testing.T) {
...
-		existingPipeline       *numaflowv1.Pipeline
-		expectedRolloutPhase   apiv1.Phase
+		name                    string
+		newISBSvcSpec           numaflowv1.InterStepBufferServiceSpec
+		existingISBSvcDef       *numaflowv1.InterStepBufferService
+		existingStatefulSetDef  *appsv1.StatefulSet
+		existingPipelineRollout *apiv1.PipelineRollout
+		existingPipeline        *numaflowv1.Pipeline
+		existingPauseRequest    *bool // was ISBServiceRollout previously requesting pause?
+		expectedPauseRequest    *bool // after reconcile(), should it be requesting pause?
```
Comment (juliev0) [inline internal/controller/isbservicerollout_controller_test.go:289]: set the current state of the ISBServiceRollout's pause request prior to the test, and verify if after

File: internal/controller/numaflowcontrollerrollout_controller_test.go [hunk 4]
```diff
@@ -310,3 +346,259 @@ func Test_getControllerDeploymentVersion(t *testing.T) {
...
+			},
+			expectedResultControllerVersion: "1.2.0",
+		},
+		{
+			name:                    "new Controller version, pipelines not yet paused",
+			newControllerVersion:    "1.2.1",
+			existingController:      createDeploymentDefinition("quay.io/numaproj/numaflow:v1.2.0", false),
+			existingPipelineRollout: createPipelineRollout(numaflowv1.PipelineSpec{InterStepBufferServiceName: defaultISBSvcRolloutName}, map[string]string{}, map[string]string{}),
+			existingPipeline:        createDefaultPipelineOfPhase(numaflowv1.PipelinePhasePausing),
+			existingPauseRequest:    &falseValue,
```
Comment (dpadhiar) [inline internal/controller/numaflowcontrollerrollout_controller_test.go:417]: For this case, pause request is false but the pipeline is in pausing state, is pause request only true when the resources are fully paused?

**Review by dpadhiar — APPROVED** [review]

File: internal/controller/numaflowcontrollerrollout_controller_test.go (hunk 4 above)
Comment (juliev0) [reply internal/controller/numaflowcontrollerrollout_controller_test.go:417]: Good catch. If it's only the Numaflow Controller Rollout which is the reason for the Pipeline being in a Pausing state, then the `existingPauseRequest` that the Numaflow Controller Rollout has made should in fact be set to true. 

But technically the Pipeline could be pausing if either the Pipeline or the ISBService is in the middle of an update.

Comment (juliev0) [reply internal/controller/numaflowcontrollerrollout_controller_test.go:417]: Perhaps the test would have been easier to read and make sense of if I made the existing Pipeline in a "Running" state, though.

Comment (dpadhiar) [reply internal/controller/numaflowcontrollerrollout_controller_test.go:417]: No worries, that makes sense overall

**Pull request merged**